File: go/api/arch/member.go
/**
* Copyright (c) 2015-2022, FineX, All Rights Reserved.
* @author Thomas.Zhuang
* @date 2022/06/13
 */

package arch

import (
	"api-demo/base"
	"api-demo/model"
	"encoding/json"
	"fmt"
)

// MemberApiClient 成员Api
type MemberApiClient struct {
	*base.ApiClient
}

func (api *MemberApiClient) ValidVersion(version string) bool {
	return map[string]bool{
		"v1": true,
		"v2": true,
	}[version]
}

func (api *MemberApiClient) DefaultVersion() string {
	return "v2"
}

// DoRequest 发送http请求
func (api *MemberApiClient) DoRequest(option *base.RequestOption) (responseBody []byte, err error) {
	if !api.ValidVersion(api.Version) {
		api.Version = api.DefaultVersion()
	}
	return api.ApiClient.DoRequest(option)
}

// DeptMemberList 获取部门成员（递归）
func (api *MemberApiClient) DeptMemberList(deptNo int, options map[string]interface{}) (responseBody *model.MemberList, err error) {
	payload, err := json.Marshal(options)
	if err != nil {
		return
	}

	response, err := api.DoRequest(&base.RequestOption{
		Method:  "POST",
		Path:    fmt.Sprintf("department/%d/member_list", deptNo),
		Payload: payload,
	})
	if err != nil {
		return
	}

	err = json.Unmarshal(response, &responseBody)
	return
}

// UserInfo 获取成员
func (api *MemberApiClient) UserInfo(username string) (responseBody *model.MemberResponse, err error) {
	response, err := api.DoRequest(&base.RequestOption{
		Method: "POST",
		Path:   fmt.Sprintf("user/%s/user_retrieve", username),
	})
	if err != nil {
		return
	}

	err = json.Unmarshal(response, &responseBody)
	return
}

// UserCreate 创建成员
func (api *MemberApiClient) UserCreate(name string, options map[string]interface{}) (responseBody *model.MemberResponse, err error) {
	options["name"] = name
	payload, err := json.Marshal(options)
	if err != nil {
		return
	}

	response, err := api.DoRequest(&base.RequestOption{
		Method:  "POST",
		Path:    "user/create",
		Payload: payload,
	})
	if err != nil {
		return
	}

	err = json.Unmarshal(response, &responseBody)
	return
}

// UserUpdate 更新成员
func (api *MemberApiClient) UserUpdate(username string, options map[string]interface{}) (responseBody *model.MemberResponse, err error) {
	payload, err := json.Marshal(options)
	if err != nil {
		return
	}

	response, err := api.DoRequest(&base.RequestOption{
		Method:  "POST",
		Path:    fmt.Sprintf("user/%s/update", username),
		Payload: payload,
	})
	if err != nil {
		return
	}

	err = json.Unmarshal(response, &responseBody)
	return
}

// UserDelete 删除成员
func (api *MemberApiClient) UserDelete(username string) (responseBody *model.Response, err error) {
	response, err := api.DoRequest(&base.RequestOption{
		Method: "POST",
		Path:   fmt.Sprintf("user/%s/delete", username),
	})
	if err != nil {
		return
	}

	err = json.Unmarshal(response, &responseBody)
	return
}

// UserBatchDelete 批量删除成员
func (api *MemberApiClient) UserBatchDelete(usernames []string) (responseBody *model.Response, err error) {
	payload, err := json.Marshal(map[string]interface{}{
		"usernames": usernames,
	})
	if err != nil {
		return
	}

	response, err := api.DoRequest(&base.RequestOption{
		Method:  "POST",
		Path:    "user/batch_delete",
		Payload: payload,
	})
	if err != nil {
		return
	}

	err = json.Unmarshal(response, &responseBody)
	return
}

// UserImport 批量导入成员
func (api *MemberApiClient) UserImport(users []model.Member) (responseBody *model.Response, err error) {
	payload, err := json.Marshal(map[string]interface{}{
		"users": users,
	})
	if err != nil {
		return
	}

	response, err := api.DoRequest(&base.RequestOption{
		Method:  "POST",
		Path:    "user/import",
		Payload: payload,
	})
	if err != nil {
		return
	}

	err = json.Unmarshal(response, &responseBody)
	return
}

File: python/src/api/arch/member.py
from src.model.http.api_client import ApiClient
from ...constants.http_constant import HttpConstant
from ...model.http.http_request_param import HttpRequestParam

# 合法的版本
valid_versions = ('v5')
# 默认版本
default_version = 'v5'


class MemberApiClient(ApiClient):

    def __init__(self, api_key, host):
        ApiClient.__init__(self, api_key, host, valid_versions, default_version)

    """
     获取部门成员（递归）
     Arguments:
        dept_no: int 部门编号
        hasChild: boolean 是否递归获取
        version: 版本
    """

    def deptMemberList(self, dept_no, has_child, version=default_version):
        path = HttpConstant.DEPT_PATH.format(suffix='user/list',
                                             version=self.getValidVersion(version))
        request_param = HttpRequestParam(path, {"has_child": has_child, 'dept_no': dept_no})
        return self.send_post(request_param)

    """
    创建 成员
    Arguments:
        user_create_param: UserCreateParam class 对象
        version: 版本
    """

    def userCreate(self, user_create_param, version=default_version):
        path = HttpConstant.MEMBER_PATH.format(suffix='create', version=self.getValidVersion(version))
        request_param = HttpRequestParam(path, user_create_param)
        return self.send_post(request_param)

    """
    根据成员名称 获取成员
    Arguments:
        user_name: string 用户名称
        version: 版本
    """

    def userInfo(self, user_name, version=default_version):
        path = HttpConstant.MEMBER_PATH.format(version=self.getValidVersion(version),
                                               suffix='get')
        request_param = HttpRequestParam(path, {'username': user_name})
        return self.send_post(request_param)

    """
    更新成员
    Arguments:
        user_name: string 用户名称
        version: 版本
    """

    def userUpdate(self, user_update_param, version=default_version):
        path = HttpConstant.MEMBER_PATH.format(version=self.getValidVersion(version), suffix='update')
        request_param = HttpRequestParam(path,
                                         {'name': user_update_param.getName(),
                                          'departments': user_update_param.getDepartments(),
                                          'username': user_update_param.getUserName()})
        return self.send_post(request_param)

    """
    删除 成员
    Arguments:
        user_name: string 用户名称
        version: 版本
    """

    def userDelete(self, user_name, version=default_version):
        path = HttpConstant.MEMBER_PATH.format(version=self.getValidVersion(version),
                                               suffix='delete')
        request_param = HttpRequestParam(path, {'username': user_name})
        return self.send_post(request_param)

    """
    批量 删除 成员
    Arguments:
        user_name_list: list string 用户名称
        version: 版本
    """

    def userBatchDelete(self, user_name_list, version=default_version):
        path = HttpConstant.MEMBER_PATH.format(suffix='batch_delete', version=self.getValidVersion(version))
        request_param = HttpRequestParam(path, {'usernames': user_name_list})
        return self.send_post(request_param)

    '''
    批量导入成员
    Arguments:
        users: list<DeptCreateParam>
        version: 版本
    '''

    def userImport(self, users, version=default_version):
        path = HttpConstant.MEMBER_PATH.format(suffix='import', version=self.getValidVersion(version))
        request_param = HttpRequestParam(path, {'users': users})
        return self.send_post(request_param)
